Common/cpp/SharedItems/Shareables.h

```diff
@@ -0,0 +1,249 @@
+  JSRuntimeHelper *runtimeHelper;
+  ShareableWorklet(
+      JSRuntimeHelper *_runtimeHelper,
+      jsi::Runtime &rt,
+      const jsi::Object &worklet)
+      : ShareableObject(rt, worklet), runtimeHelper(_runtimeHelper) {
+    valueType = WorkletType;
+  }
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    jsi::Value obj = ShareableObject::toJSValue(rt);
+    return runtimeHelper->valueUnpacker->call(rt, obj);
+  }
+};
+
+class ShareableRemoteFunction : public RetainingShareable,
+                                public std::enable_shared_from_this<ShareableRemoteFunction> {
+ public:
+  JSRuntimeHelper *runtimeHelper;
+  jsi::Function function;
+  ShareableRemoteFunction(
+      JSRuntimeHelper *runtimeHelper,
+      jsi::Runtime &rt,
+      jsi::Function &&function)
+      : RetainingShareable(rt, RemoteFunctionType),
+        runtimeHelper(runtimeHelper),
+        function(std::move(function)) {}
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    if (runtimeHelper->isUIRuntime(rt)) {
+      return ShareableJSRef::newHostObject(rt, shared_from_this());
+    } else {
+      return jsi::Value(rt, function);
+    }
+  }
+};
+
+class ShareableHandle : public RetainingShareable {
+private:
+  JSRuntimeHelper *runtimeHelper;
+  std::shared_ptr<ShareableObject> initializer;
+public:
+  ShareableHandle(JSRuntimeHelper *runtimeHelper, jsi::Runtime &rt, const jsi::Object &initializerObject) : RetainingShareable(rt, HandleType), runtimeHelper(runtimeHelper) {
+    initializer = std::make_shared<ShareableObject>(rt, initializerObject);
+  }
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    auto initObj = initializer->getJSValue(rt);
+    auto value = runtimeHelper->valueUnpacker->call(rt, initObj);
+    initializer = nullptr; // we can release ref to initializer as this method should be called at most once
+    return value;
+  }
+};
+
+class ShareableString : public Shareable {
+ public:
+  ShareableString(jsi::Runtime &rt, const jsi::String &string)
+      : Shareable(StringType) {
+    data = string.utf8(rt);
+  }
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    return jsi::String::createFromUtf8(rt, data);
+  }
+
+ protected:
+  std::string data;
+};
+
+class ShareableScalar : public Shareable {
+ public:
+  ShareableScalar(double number) : Shareable(NumberType) {
+    data.number = number;
+  };
+  ShareableScalar(bool boolean) : Shareable(BooleanType) {
+    data.boolean = boolean;
+  };
+  ShareableScalar() : Shareable(UndefinedType) {}
+  ShareableScalar(std::nullptr_t) : Shareable(NullType) {}
+
+  jsi::Value toJSValue(jsi::Runtime &rt) override {
+    switch (valueType) {
+      case Shareable::UndefinedType:
+        return jsi::Value();
+      case Shareable::NullType:
+        return jsi::Value(nullptr);
+      case Shareable::BooleanType:
+        return jsi::Value(data.boolean);
+      case Shareable::NumberType:
+        return jsi::Value(data.number);
+      default:
+        throw "invalid scalar type";
```

Comment:
```suggestion
        throw std::runtime_error("Unsupported scalar type");
```

Reply:
this actually means it is "invalid" rather than unsupported. We only create ShareableScalar objects for supported types. This error should never been reached as the constructor is only called for NumberType, BooleanType, UndefinedType and NullType however putting default section there is necessary to satisfy compiler (there are more possible value types but neither is a scalar). I changed the error message accordingly.